main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+        }
+      }
+      // INVARIANT: We apply the current substitution to all remaining equations.
+      pending = subst(rest)
+    }
+
+    // Reverse the unsolved equations to ensure they are returned in the original order.
+    (pending.reverse, subst)
+  }
+
+  /**
+    * Propagates variables through the equation system. Cannot fail.
+    *
+    * The observation is that we if we have simple equations of the form: `x ~ y`
+    * then we can create a substitution that binds `[x -> y]`. Every time we create
+    * a binding, we must be careful to update both unsolved and pending equations.
+    *
+    * For example, if the equation system is:
+    *
+    * {{{
+    *     x78914 ~ (c1498 ∧ c1500 ∧ c1501)
+    *     x78914 ~ (x78926 ∧ x78923 ∧ x78917)
+    *     x78917 ~ x127244
+    *     x78921 ~ x127251
+    *     x78923 ~ (x127249 ∧ x127247 ∧ x127248)
+    *     x78926 ~ (x127254 ∧ x127252)
+    * }}}
+    *
+    * then after variable propagation it is:
+    *
+    * {{{
+    *     x78914 ~ (c1498 ∧ c1500 ∧ c1501)
+    *     x78914 ~ (x78926 ∧ x78923 ∧ x127244)
+    *     x78923 ~ (x127249 ∧ x127247 ∧ x127248)
+    *     x78926 ~ (x127254 ∧ x127252)
+    * }}}
+    *
+    * with the substitution:
+    *
+    * {{{
+    *      x78917 -> x127244
+    *     x127251 -> x78921
+    * }}}
+    *
+    * Returns a list of unsolved equations and a partial substitution.
+    */
+  private def propagateVars(l: List[Equation], s: BoolSubstitution): (List[Equation], BoolSubstitution) = {
+    var pending = l
+    var subst = s
+
+    var unsolved: List[Equation] = Nil
+
+    // INVARIANT: The current substitution has been applied to BOTH unsolved AND pending equations.
+    while (pending.nonEmpty) {
+      // Extract the next equation and update the pending equations.
+      val e = pending.head
+      pending = pending.tail
+
+      e match {
+        case Equation(Term.Var(x), Term.Var(y), _) =>
+          // Case 1: We have found an equation: `x ~ y`.
+          // Construct a singleton substitution `[x -> y]`.
+          val singleton = BoolSubstitution.singleton(x, Term.Var(y))
+
+          // Apply the singleton substitution to the pending equations.
+          pending = singleton(pending)
+
+          // Apply the singleton substitution to the unsolved equations.
+          unsolved = singleton(unsolved)
+
+          // Update the current substitution with the new binding.
+          subst = singleton @@ subst
+
+        case _ =>
+          // Case 2: We have some other type of equation. Add it to the list of unsolved equations.
+          // The invariant ensures that the current substitution has already been applied to the equation.
+          unsolved = e :: unsolved
+      }
+    }
+
+    // Reverse the unsolved equations to ensure they are returned in the original order.
+    (unsolved.reverse, subst)
+  }
+
+  /**
+    * Assigns variables. Given a unification equation `x ~ t` we can assign `[x -> t]` if `x` does not occur in `t`.
```

Comment:
This step includes the behavior or variable propagation, no?

Reply:
Yes and no. There is no fixpoint here.

Reply:
If your point is that this function "does the same work" as varPropagation then notice that this is true for __all phases__ because SVE can solve any equation.